src/webgpu/util/f32_interval.ts

```diff
@@ -1676,6 +1708,51 @@ export function reflectInterval(x: number[], y: number[]): F32Vector {
   return runVectorPairToVectorOp(toF32Vector(x), toF32Vector(y), ReflectIntervalOp);
 }
 
+/**
+ * Calculate acceptance interval vectors of reflect(i, s, r)
+ *
+ * refract is a singular function in the sense that it is the only builtin that
+ * takes in (F32Vector, F32Vector, F32) and returns F32Vector and is basically
+ * defined in terms of other functions.
+ *
+ * Instead of implementing all of the framework code to integrate it with its
+ * own operation type/etc, it instead has a bespoke implementation that is a
+ * composition of other builtin functions that use the framework.
+ */
+export function refractInterval(i: number[], s: number[], r: number): F32Vector {
+  assert(
+    i.length === s.length,
+    `refract is only defined for vectors with the same number of elements`
+  );
+
+  const r_squared = multiplicationInterval(r, r);
+  const dot = dotInterval(s, i);
+  const dot_squared = multiplicationInterval(dot, dot);
+  const one_minus_dot_squared = subtractionInterval(1, dot_squared);
+  const k = subtractionInterval(1.0, multiplicationInterval(r_squared, one_minus_dot_squared));
+
+  if (k.containsSubnormals()) {
```

Comment:
I had to look up this function as it's a little non-obvious that it would return true for zero.

Reply:
Changing the name to `containsZeroOrSubnormals`, since the better describes the check.